src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,160 @@
+
+        T[] newList = (T[]) new Object[(int) (arrayList.length)];
+        // adding at particular position is made in a few steps
+        // step 1
+        int startSrc = 0;
+        int startDest = 0;
+        int count = index;
+        System.arraycopy(arrayList, startSrc, newList, startDest, count);
+        newList[index] = value;
 
+        // step 2
+        startSrc = index;
+        startDest += count + 1;
+        count = size - count;
+        System.arraycopy(arrayList, startSrc, newList, startDest, count);
+
+        arrayList = newList;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return arrayList[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        arrayList[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T[] newList = (T[]) new Object[(int) (arrayList.length)];
+        // extraction is made in two steps
+        // step 1
+        int startSrc = 0;
+        int startDest = 0;
+        int count = index;
+        System.arraycopy(arrayList, startSrc, newList, startDest, count);
+
+        // step 2
+        startSrc = index + 1;
+        startDest += count;
+        count = size - count - 1;
+        System.arraycopy(arrayList, startSrc, newList, startDest, count);
+
+        T removed = get(index);
+        arrayList = newList;
+        size--;
+
+        return removed;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getIndex(element);
+        if (index == -1) {
+            throw new NoSuchElementException("There is no such element in the list!");
+        }
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return (size == 0);
+    }
+
+    @Override
+    public String toString() {
+
+        StringBuilder b = new StringBuilder();
+        b.append('[');
+        for (int i = 0; i < size; i++) {
+            b.append(String.valueOf(arrayList[i]));
+            if (i == (size - 1)) {
+                b.append(']');
+            } else {
+                b.append(", ");
+            }
+        }
+
+        return "ArrayList = " + b.toString();
+    }
+
+    private void checkIndex(int index) {
+        if (index >= 0 && index < size) {
+            return;
+        }
+        throw new ArrayListIndexOutOfBoundsException("IndexOutOfBounds: " + index
+                + " is out of the ranges");
+    }
+
+    private void resize(int newSize) {
+        if (newSize <= arrayList.length) {
+            return;
+        }
```

Comment:
```suggestion
    private void resize() {
        if (size < arrayList.length) {
            return;
        }
```
тобі не потрібно передавати на вхід параметри, ти вже маєш до них доступ